front: fix track offset nodes in NGE

emersion opened the pull request:
When building an OSRD train path, we were unconditionally using the trigram from the NGE node's betriebspunktName field. This worked fine for OPs with a trigram, but fell apart for track offsets (and infrastructures without trigrams).

Instead, reconstruct the PathItemLocation from the path_item_key stored in MacroEditorState.nodes.

As a bonus, this makes it so path item location types are no longer changed to trigrams when editing a train from NGE: if a train was created with UICs from OSRD, editing it in NGE will retain UICs now.

We still need to keep the old betriebspunktName-based logic for NGE JSON file imports.

Closes: https://github.com/OpenRailAssociation/osrd/issues/12618

emersion committed:
front: fix track offset nodes in NGE

When building an OSRD train path, we were unconditionally using
the trigram from the NGE node's betriebspunktName field. This worked
fine for OPs with a trigram, but fell apart for track offsets (and
infrastructures without trigrams).

Instead, reconstruct the PathItemLocation from the path_item_key
stored in MacroEditorState.nodes.

As a bonus, this makes it so path item location types are no
longer changed to trigrams when editing a train from NGE: if a
train was created with UICs from OSRD, editing it in NGE will
retain UICs now.

We still need to keep the old betriebspunktName-based logic for
NGE JSON file imports.

Signed-off-by: Simon Ser <contact@emersion.fr>
Closes: https://github.com/OpenRailAssociation/osrd/issues/12618

Caracol3 commented on front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/MacroEditorState.ts:
Can't we have a problem if `op.extensions?.sncf` and `op.extensions?.identifier` are null and then pass invalid keys like `trigram:undefined` and `uic:undefined` to `parsePathKey()` ?

emersion replied:
I don't think this can happen because `parsePathKey()` only consumes keys generated by `getPathKey()` (as opposed to `getPathKeys()`, plural). Nevertheless, you make a good point even if it's orthogonal to this PR: the UIC and trigram are optional OP fields so I think it makes sense to avoid generating bogus path keys when they are missing. I've pushed a commit to fix that!

Caracol3 replied:
You're right and  thanks for the modif

emersion committed:
front: handle missing UIC/trigram in MacroEditorState.getPathKeys()

These are optional fields so they can be undefined. Avoid generating
bogus path item keys such as "uic:undefined" in that case.

Signed-off-by: Simon Ser <contact@emersion.fr>

Caracol3 reviewed (approved):
LGTM (tested)

Math-R reviewed (approved):
lgtm tested
